### ISSUE TITLE
False positive when working with ArrayAccess

### ISSUE THREAD

**Issue by BladeMF** [phpstan/phpstan-doctrine#557]
[This code](https://phpstan.org/r/9e8d0c74-3760-4a9c-884e-ec0fd6b27cb9) fails with the error `... expects ArrayAccess<int, covariant object>, Doctrine\Common\Collections\ArrayCollection<int, stdClass> given.`

**Comment by ondrejmirtes** [comment]
You should be able to reproduce that problem in the playground, not sure why you can't.

**Comment by BladeMF** [comment]
Is that extension installed there?

**Comment by ondrejmirtes** [comment]
AFAIK there is nothing special in the extension you can't reproduce on the playground. Just stub files.

**Comment by BladeMF** [comment]
But if I remove the extension the error goes away. Would you like me to test something specific? I can try and dig in deeper?

**Comment by ondrejmirtes** [comment]
What if you only include the related stub files from extension.neon here?

If you're not successful in reproducing the issue on the playground, you should be able to reproduce it in a small repository where you can include the phpstan-doctrine extension.

**Comment by BladeMF** [comment]
I found it, gimme a second to find the stub and link it.

**Comment by BladeMF** [comment]
This `|null` causes it: https://github.com/phpstan/phpstan-doctrine/blob/0871900872abde0d93e4bbd8c681a9985bcd927e/stubs/Collections/Collection.stub#L13

**Comment by ondrejmirtes** [comment]
Can you reproduce it on the playground now?

**Comment by BladeMF** [comment]
This is the original `Collection`:
```php
/**
 * @psalm-template TKey of array-key
 * @psalm-template T
 * @template-extends ReadableCollection<TKey, T>
 * @template-extends ArrayAccess<TKey, T>
 */
interface Collection extends ReadableCollection, ArrayAccess
```

this is the stub:
```php
/**
 * @template TKey of array-key
 * @template T
 * @extends IteratorAggregate<TKey, T>
 * @extends ArrayAccess<TKey|null, T>
 * @extends ReadableCollection<TKey, T>
 */
interface Collection extends Countable, IteratorAggregate, ArrayAccess, ReadableCollection
```

**Comment by BladeMF** [comment]
Also, `Countable, IteratorAggregate` is redundant here as `ReadableCollection` extends them.

**Comment by ondrejmirtes** [comment]
Not on older Doctrine versions.

**Comment by BladeMF** [comment]
Touche. Is the null there as well?

**Comment by ondrejmirtes** [comment]
Please reproduce the problem on the playground or in a small repository so I can evaluate whether the problem is valid or not.

**Comment by BladeMF** [comment]
https://phpstan.org/r/ce3ebe06-1a2c-4e3c-a282-77e79ef43937

**Comment by ondrejmirtes** [comment]
Summoning our variance expert @jiripudil here: what do you think, is this a legit errror? https://phpstan.org/r/ce3ebe06-1a2c-4e3c-a282-77e79ef43937

Thanks 😊

**Comment by ondrejmirtes** [comment]
Without |null it disappears which is probably expected too: https://phpstan.org/r/9df61dda-ada2-42fe-bb10-ff9292d21969

The question is - do we need |null in the stub or not?

**Comment by jiripudil** [comment]
Hi, I believe this to be legit. ArrayAccess [is invariant](https://github.com/phpstan/phpstan-src/blob/8b7adac5c60138ce79d383c7fc3a9167f7c942c1/stubs/ArrayObject.stub#L4) in its key, and therefore `ArrayAccess<int|null, *>` and `ArrayAccess<int, *>` are two different, unrelated types.

**Comment by BladeMF** [comment]
You are both correct. The error in isolation is correct, it is _not a false positive_. 

The trouble is that in Doctrine 2.8 `null` is not allowed as a key - i.e. it's declared as just `ArrayAccess<array-key, *>`, and `array-key` is defined as `array-key is the supertype (but not a union) of int and string.`. I don't mind changing my code really, but imagine how deep I had to dig to understand why does Phpstan give me an error in a place where the code says it should not. I imagine there are other people like that as well. 

My suggestion is that a new version of this package is created, requiring Doctrine (whichever version started to use `array-key`), that fixes the interface inheritance on `Collection` and removes the `|null` and so people with the old version can stay on the old definitions and people like me can use the new version.

My specific use case (just for the record) is a package that can work with both arrays and Doctrine collections without having dependency on Doctrine specifically (so I can't use`array|Collection`, for example).

**Comment by ondrejmirtes** [comment]
I've removed the `|null` part, we'll see what happens: https://github.com/phpstan/phpstan-doctrine/commit/1667bda042320163fe503dd15e9fee3d37cb31cf

**Comment by ondrejmirtes** [comment]
Context: https://github.com/psalm/psalm-plugin-doctrine/pull/67 (Psalm might need `|null` but we don't).
